### PR TITLE
feat: FromStr derive could support setting the error type

### LINKED ISSUE #91: FromStr derive could support setting the error type
When you derive on an enum, sometimes you want to implement some `From<XXXNotFound>` for some other error types, with maybe more explicit messages. It would be great if we could have an attribute that enables us to set the function that would be used to build the error from the original `&str`.

e.g.
```rust
#[derive(EnumString)]
#[strum(parse_err_ty = "SomeEnumNotFound", parse_err_fn = "some_enum_not_found_err")]
pub enum SomeEnum {
	A,
	B,
}

pub struct SomeEnumNotFound;
fn some_enum_not_found_err(_not_found: &str) -> SomeEnumNotFound { SomeEnumNotFound }
```

generating

```rust
impl FromStr for SomeEnum {
    type Err = SomeEnumNotFound;
    fn from_str(s: &str) -> Result<SomeEnum, Self::Err> {
        match s {
            "A" => A,
            "B" => B,
            other => some_enum_not_found_err(other),
        }
    }
}
```

If `parse_err_fn` is not specified but `parse_err_ty` is, we could fallback on `From<strum::ParseError>`.

### PR DESCRIPTION
**Pull request by JimChenWYU**
resolve #91 #332 #221

PR checker failed because #381

### CODE REVIEW

**Comment by Peternator7** [comment]
This looks really good! Sorry for the delay reviewing it

File: strum_macros/src/helpers/mod.rs [hunk 1]
```diff
@@ -13,6 +13,13 @@ use proc_macro2::Span;
 use quote::ToTokens;
 use syn::spanned::Spanned;
 
+pub fn missing_parse_err_attr_error() -> syn::Error {
+    syn::Error::new(
+        Span::call_site(),
+        "`parse_err_ty` and `parse_err_fn` attribute is both required.",
```
Comment (martin-g) [inline strum_macros/src/helpers/mod.rs:19]:
```suggestion
        "`parse_err_ty` and `parse_err_fn` attributes are both required.",
```

**Comment by sebschrader** [comment]
Please reconsider using string literals for paths. Attributes nowadays support arbitrary expressions (including paths) instead of only string literals.

**Comment by JimChenWYU** [comment]
> Please reconsider using string literals for paths. Attributes nowadays support arbitrary expressions (including paths) instead of only string literals.
> 
> One of the authors of `rust-analyzer` recently highlighted that using string literals for such things is very bad for IDE support, in his EuroRust [2024 presentation](https://eurorust.eu/2024/talks/powerful-macros-good-ide-support-no-headaches-choose-two/).

So you want like this?

```rust
#[derive(Debug, EnumString)]
#[strum(
    parse_err_fn = crate::custom::some_enum_not_found_err,
    parse_err_ty =crate::custom::CaseCustomParseErrorNotFoundError
)]
enum CaseCustomParseErrorEnum {
    #[strum(serialize = "red")]
    Red,
    #[strum(serialize = "blue")]
    Blue,
}

mod custom {
    #[derive(Debug, Clone, Eq, PartialEq, Hash)]
    struct CaseCustomParseErrorNotFoundError(String);
    impl std::fmt::Display for CaseCustomParseErrorNotFoundError {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            writeln!(f, "not found `{}`", self.0)
        }
    }
    impl std::error::Error for CaseCustomParseErrorNotFoundError {}
    pub fn some_enum_not_found_err(s: &str) -> CaseCustomParseErrorNotFoundError {
        CaseCustomParseErrorNotFoundError(s.to_string())
    }
}
```

**Comment by Peternator7** [comment]
@sebschrader, I think it's a reasonable point. I've opened #410 to change this behavior since it's not released yet.

**Pull request merged**